llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp

```diff
@@ -1240,6 +1269,116 @@ struct FunctionStackPoisoner : public InstVisitor<FunctionStackPoisoner> {
 
 } // end anonymous namespace
 
+// Append a new argument "launch_data" to user's spir_kernels
+static void ExtendSpirKernelArgs(Module &M, FunctionAnalysisManager &FAM) {
+  SmallVector<Function *> SpirFixupFuncs;
+  for (Function &F : M) {
+    if (F.getCallingConv() == CallingConv::SPIR_KERNEL) {
+      SpirFixupFuncs.emplace_back(&F);
+    }
+  }
+
+  SmallVector<std::pair<Function *, Function *>> SpirFuncs;
+  const int LongSize = M.getDataLayout().getPointerSizeInBits();
+  auto *IntptrTy = Type::getIntNTy(M.getContext(), LongSize);
+
+  for (auto *F : SpirFixupFuncs) {
+    SmallVector<Type *, 16> Types;
+    for (Function::const_arg_iterator I = F->arg_begin(), E = F->arg_end();
+         I != E; ++I) {
+      Types.push_back(I->getType());
+    }
+
+    // New argument type: uintptr_t as(1)*, as it's allocated in USM buffer, and
+    // it can also be treated as a pointer point to the base address of private
+    // shadow memory
+    Types.push_back(IntptrTy->getPointerTo(kSpirOffloadGlobalAS));
+
+    FunctionType *NewFTy = FunctionType::get(F->getReturnType(), Types, false);
+
+    std::string OrigFuncName = F->getName().str();
+    F->setName(OrigFuncName + "_del");
```

Comment:
You can use [`Value::takeName`](https://llvm.org/doxygen/classllvm_1_1Value.html#ae855357b6c5e6e7ed1869272708a3a84) to simplify this down to  `NewF->takeName(F)`

Reply:
I think adding "_del" suffix in old func is helpful for debugging. If I forget to replace the old function in other instructions/metadata, I can know this immediately.